library/std/src/sys/pal/unix/locks/queue_rwlock.rs

```diff
@@ -0,0 +1,555 @@
+//! created. This is done at the same time as finding the tail, and thus a set
+//! tail field indicates the remaining portion of the queue is initialized.
+//!
+//! TLDR: Here's a diagram of what the queue looks like:
+//!
+//! ```text
+//! state
+//!   │
+//!   ▼
+//! ╭───────╮ next ╭───────╮ next ╭───────╮ next ╭───────╮
+//! │       ├─────►│       ├─────►│       ├─────►│ count │
+//! │       │      │       │      │       │      │       │
+//! │       │      │       │◄─────┤       │◄─────┤       │
+//! ╰───────╯      ╰───────╯ prev ╰───────╯ prev ╰───────╯
+//!                      │                           ▲
+//!                      └───────────────────────────┘
+//!                                  tail
+//! ```
+//!
+//! Invariants:
+//! 1. At least one node must contain a non-null, current `tail` field.
+//! 2. The first non-null `tail` field must be valid and current.
+//! 3. All nodes preceding this node must have a correct, non-null `next` field.
+//! 4. All nodes following this node must have a correct, non-null `prev` field.
+//!
+//! Access to the queue is controlled by the `QUEUE_LOCKED` bit, which threads
+//! try to set both after enqueuing themselves to eagerly add backlinks to the
+//! queue and after unlocking the lock to wake the next waiter(s). This is done
+//! atomically at the same time as the enqueuing/unlocking operation. The thread
+//! releasing the `QUEUE_LOCK` bit will check the state of the lock and wake up
+//! waiters as appropriate. This guarantees forward-progress even if the unlocking
+//! thread could not acquire the queue lock.
+//!
+//! ## Memory orderings
+//!
+//! To properly synchronize changes to the data protected by the lock, the lock
+//! is acquired and released with [`Acquire`] and [`Release`] ordering, respectively.
+//! To propagate the initialization of nodes, changes to the queue lock are also
+//! performed using these orderings.
+
+#![forbid(unsafe_op_in_unsafe_fn)]
+
+use crate::cell::OnceCell;
+use crate::hint::spin_loop;
+use crate::mem;
+use crate::ptr::{self, invalid_mut, null_mut, NonNull};
+use crate::sync::atomic::{
+    AtomicBool, AtomicPtr,
+    Ordering::{AcqRel, Acquire, Relaxed, Release},
+};
+use crate::sys_common::thread_info;
+use crate::thread::Thread;
+
+// Locking uses exponential backoff. `SPIN_COUNT` indicates how many times the
+// locking operation will be retried.
+// `spin_loop` will be called `2.pow(SPIN_COUNT) - 1` times.
+const SPIN_COUNT: usize = 7;
+
+type State = *mut ();
+type AtomicState = AtomicPtr<()>;
+
+const UNLOCKED: State = invalid_mut(0);
+const LOCKED: usize = 1;
+const QUEUED: usize = 2;
+const QUEUE_LOCKED: usize = 4;
+const SINGLE: usize = 8;
+const MASK: usize = !(QUEUE_LOCKED | QUEUED | LOCKED);
+
+/// Returns a closure that changes the state to the lock state corresponding to
+/// the lock mode indicated in `write`.
+#[inline]
+fn lock(write: bool) -> impl Fn(State) -> Option<State> {
```

Comment:
Does this need to be a function-returning function? It's only used in two places, one of which as `lock(false)`, right? What about just having two simple functions, as follows?

```rust
fn write_lock(State) -> Option<State>;
fn read_lock(State) -> Option<State>;
```

Reply:
I think this one is cleanest, otherwise we end up with either function pointers or a
```rust
if let Some(next) = if write { write_lock(state) } else { read_lock(state) } {
```
inside `lock_contended`. But I don't feel strongly about this...

Reply:
Not just this?
```rust
let update = if write { write_lock } else { read_lock };
```

Reply:
I guess that's what you mean with function pointers. That's a performance issue, I assume?

Reply:
Though that can be avoided with:
```rust
let update = |state| if write { write_lock(state) } else { read_lock(state) };
```

Reply:
> I guess that's what you mean with function pointers. That's a performance issue, I assume?

Probably not, this is all in the same function. I've changed it now.